### PR TITLE
Add buttonTagHelper and submitTagHelper for formaction

### PR DESCRIPTION
**Pull request by ctyar**
Addresses #1668

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/Resources.resx [hunk 1]
```diff
@@ -147,4 +147,16 @@
   <data name="PropertyOfTypeCannotBeNull" xml:space="preserve">
     <value>The '{0}' property of '{1}' must not be null.</value>
   </data>
+  <data name="ButtonTagHelper_CannotOverrideFormAction" xml:space="preserve">
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/Resources.resx]: Suggest renaming existing resources e.g. `AnchorTagHelper_CannotOverrideHref` and reusing them here. Resources are `internal` so there's no compat worry. And maintenance will be easier if we keep the numbers and repetitions down.

File: test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/ButtonTagHelperTest.cs [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+
+            var urlHelperFactory = new Mock<IUrlHelperFactory>();
+            urlHelperFactory
+                .Setup(f => f.GetUrlHelper(It.IsAny<ActionContext>()))
+                .Returns(urlHelper.Object);
+
+            var htmlGenerator = new TestableHtmlGenerator(metadataProvider, urlHelper.Object);
+            var viewContext = TestableHtmlGenerator.GetViewContext(model: null,
+                                                                   htmlGenerator: htmlGenerator,
+                                                                   metadataProvider: metadataProvider);
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/ButtonTagHelperTest.cs]: Put the first argument on its own line too and don't get too fancy w/ indentation:
``` c#
var viewContext = TestableHtmlGenerator.GetViewContext(
    model: null,
    htmlGenerator: htmlGenerator,
    metadataProvider: metadataProvider);
```

**Comment by dougbu** [comment]
I haven't looked at the tests in detail. But the general direction is very solid.

One question: The `<a>` and `<form>` tag helpers already duplicate each other a fair amount. The new `<button>` and `<submit>` helpers both look almost identical to the `<a>` one. Could we refactor things e.g. to a static class? That way, future fixes will be done everywhere and we won't miss a tag helper or three? (Can't change the existing tag helpers to have a base class because that would be a breaking change and @rynowak would hate me forever more.)

**Comment by Eilon** [comment]
@dougbu assigning this to you since you already started looking at it.

@ctyar - sorry for the delay getting back to you. Do you have thoughts on @dougbu 's comments?

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/Resources.resx (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/Resources.resx]: :see_no_evil: Never mind. "a" versus "an" and the different ordering of the `FormTagHelper_CannotOverrideAction` parameters makes this difficult.

**Comment by dougbu** [comment]
@ctyar thanks very much for your submission. Please see #5201 for the direction we're taking this work. Your commit remains and will be part of the final push to GitHub of course.

**Comment by ctyar** [comment]
@dougbu Sorry that couldn't finish the work, I hope this does not cause any inconvenience to you guys

**Pull request closed without merging**